Add categorization of AARch64 instruction mix

kuhanov opened the pull request:
Added categorization for AARch64 instruction mix.
Each opcode is set to categories from the following list:
 - Integer arithmetic operations
 - Floating-Point arithmetic operations
 - Loads 
 - Stores
 - Branches
 - Operations with vector registers (SIMD) 
 - Other types of instructions

Calculation of category is based on C4.1.67 of ARM V8 Architecture reference manual
     *  https://developer.arm.com/documentation/ddi0487/

Also, we could add our categorization for x86_64 for issue #5994. Unfortunantly I don't have access to mentioned file https://source.corp.google.com/piper///depot/google3/third_party/dynamorio/trunk/core/ir/instr_api.h;l=1816?q=instr_api.h to check it.

Thx, Kirill

derekbruening commented:
> Also, we could add our categorization for x86_64 for issue #5994. Unfortunantly I don't have access to mentioned file https://source.corp.google.com/piper///depot/google3/third_party/dynamorio/trunk/core/ir/instr_api.h;l=1816?q=instr_api.h to check it.

Bad link: updated it to https://github.com/DynamoRIO/dynamorio/blob/master/core/ir/instr_api.h#L1845, just points at instr_is_floating().  There is also instr_is_floating_ex() with dr_fp_type_t: https://dynamorio.org/dr__ir__instr_8h.html#a6a279ef5b19e7c4f82f8804d9490da54

There is also a proposal to provide categories of opcodes for https://dynamorio.org/google_workload_traces.html which may overlap with an API like this; not yet filed in the public tracker -- will do so.

derekbruening commented:
> Calculation of category is based on C4.1.67 of ARM V8 Architecture reference manual * https://developer.arm.com/documentation/ddi0487/

Is that an older version?  The latest PDF at that link has "C4.1.67 SVE2 Histogram Computation - Segment" so it must be a different section in your version: what is the title?

derekbruening commented on core/ir/category_api.h:
```diff
@@ -0,0 +1,15 @@
+    DR_INSTR_CATEGORY_STORE = 0x8,         /* Stores */
+    DR_INSTR_CATEGORY_BRANCH = 0x10,       /* Branches */
+    DR_INSTR_CATEGORY_SIMD = 0x20,         /* Operations with vector registers (SIMD) */
+    DR_INSTR_CATEGORY_OTHER = 0x40         /* Other types of instructions */
```
Xref #6238 where we also had a list of categories, for the Google workload traces where we can't give out opcodes (but we'd want a programmatic way to find those categories from a real opcode, so this PR would fit the bill, if on all arches).

Comparing the category lists: looks like yours is also composable which is good.  We wanted to distinguish single and double precision floating point, and to separate out FMA (fused multiply add).  Maybe those are the main differences between the two lists; the extra categories in #6238 are less important.

kuhanov replied:
> Xref #6238 where we also had a list of categories, for the Google workload traces where we can't give out opcodes (but we'd want a programmatic way to find those categories from a real opcode, so this PR would fit the bill, if on all arches).
> 
> Comparing the category lists: looks like yours is also composable which is good. We wanted to distinguish single and double precision floating point, and to separate out FMA (fused multiply add). Maybe those are the main differences between the two lists; the extra categories in #6238 are less important.

Could not find something like this for x86_64. We create just flat list for opcodes about their categories.
We could start from this list and add new ones categories. For example, some instructions could be load ans SIMD.
So, float could be in DR_INSTR_CATEGORY_MATH_FLOAT |  DR_INSTR_CATEGORY_MATH_SINGLE_FLOAT or  DR_INSTR_CATEGORY_MATH_FLOAT |  DR_INSTR_CATEGORY_MATH_DOUBLE_FLOAT

derekbruening replied:
There should be combinations today: a scatter instruction should be STORE|SIMD and gather LOAD|SIMD.  If that's not simple through the encoding bit scheme, maybe these would have to be flags added to every opcode in the codec*.txt files.  That is likely how it would be done for x86 and AArch32 as well: flags on every encoding entry (or else it has a be a big external switch statement on opcodes like today's instr_is_floating_ex()).

kuhanov replied:
That was done already over bits
        category =
            (BITS(enc, 22, 22) == 0) ? DR_INSTR_CATEGORY_STORE : DR_INSTR_CATEGORY_LOAD;
        if ((op0 & 0xc) == 0 && BITS(enc, 26, 26) == 1)
            category |= DR_INSTR_CATEGORY_SIMD;

derekbruening replied:
OK.  Sure, let's start with this, and if we want to add single vs double, FMA, atomics, memory barriers, etc. later they can be appended.

kuhanov commented:
> > Calculation of category is based on C4.1.67 of ARM V8 Architecture reference manual * https://developer.arm.com/documentation/ddi0487/
> 
> Is that an older version? The latest PDF at that link has "C4.1.67 SVE2 Histogram Computation - Segment" so it must be a different section in your version: what is the title?

Need just use full 4.1 section
C4.1 A64 instruction set encoding

kuhanov commented:
Hi, @derekbruening . Could you tell me if I need to make anything else? Could I merge this patch?
Thx, Kirill

derekbruening commented on core/ir/instr_api.h:
```diff
@@ -289,6 +289,8 @@ struct _instr_t {
 
     uint opcode;
 
+    uint category;
```
The instr_t size is tightly controlled and tuned for in the heap allocation buckets.  Adding a field requires making sure the BLOCK_SIZES in core/heap are sane, updating the commented sizes, and as a binary compatibility breakage this requires a major version bump and a listing in the release notes.

One alternative is not store anything and encode on demand when the category is asked about, like in instr_zeroes_ymmh() and others where we get the encoding info.  That has some downsides though of not having a target PC for PC-relative encoding differences and possible overhead: xref https://github.com/dvyukov/dynamorio/pull/1.

Another alternative is to fit the bits in an existing field like prefixes which is where the predicate bits are: but while they may well fit on AArch64 I think x86 uses a lot of those bits for internal decoding even if none are exposed.

derekbruening commented on core/ir/category_api.h:
```diff
@@ -0,0 +1,15 @@
+#ifndef _DR_IR_CATEGORIES_H_
+#define _DR_IR_CATEGORIES_H_ 1
+
+typedef enum _dr_instr_category_type_t {
```
Missing Doxygen docs.

derekbruening commented on core/ir/category_api.h:
```diff
@@ -0,0 +1,15 @@
+#ifndef _DR_IR_CATEGORIES_H_
```
Missing copyright and license header.

derekbruening commented on core/ir/category_api.h:
```diff
@@ -0,0 +1,15 @@
+#define _DR_IR_CATEGORIES_H_ 1
+
+typedef enum _dr_instr_category_type_t {
+    DR_INSTR_CATEGORY_UNCATEGORIZED = 0x0, /* Uncategorised. */
```
Missing Doxygen docs: here and below.

derekbruening commented on core/ir/instr_api.h:
```diff
@@ -738,6 +745,13 @@ DR_API
 instr_set_opcode(instr_t *instr, int opcode);
 
+DR_API
+/** Assumes \p category is an set of DR_INSTR_CATEGORY_ constants
```
style: text doesn't go on the `/**` line for a multi-line comment

derekbruening replied:
grammar: a set

derekbruening commented on core/ir/instr_api.h:
```diff
@@ -722,6 +724,11 @@ DR_API
 instr_get_opcode(instr_t *instr);
 
+DR_API
+/** Returns \p instr's set of categories (set of DR_INSTR_CATEGORY_ constants). */
```
Always make references to other types actual links in Doxygen comments: `#dr_instr_category_type_t`

derekbruening replied:
The docs need to explain that this is only implemented for AArch64.

derekbruening replied:
Continuing the discussion of this PR only setting categories on decode: what if I synthesize a new instruction and then ask about its categories?  The existing IR behavior would be to encode and figure things out from the encoding: that's how instr_length(), instr_zeroes_ymmh(), etc. all operate.  I would expect the same here.  At the very least this behavior of only being valid for unmodified-since-decode instructions would have to be clearly documented: but that route does not seem very desirable.

derekbruening commented on core/lib/globals_api.h:
```diff
@@ -377,9 +377,9 @@ typedef struct {
 typedef struct {
 #    ifdef X64
-    uint black_box_uint[28];
+    uint black_box_uint[30];
```
2 more?  Because of padding?  Please try to move the field to eliminate the extra uint of space.

derekbruening commented on core/lib/globals_api.h:
```diff
@@ -377,9 +377,9 @@ typedef struct {
+    uint black_box_uint[19];
 #    endif
 } instr_t;
 #else
```
I don't see any tests: all new features need tests.

derekbruening replied:
All new API routines get a listing in the changelog in release.dox.

derekbruening commented on core/ir/category_api.h:
```diff
@@ -0,0 +1,15 @@
+    DR_INSTR_CATEGORY_STORE = 0x8,         /* Stores */
+    DR_INSTR_CATEGORY_BRANCH = 0x10,       /* Branches */
+    DR_INSTR_CATEGORY_SIMD = 0x20,         /* Operations with vector registers (SIMD) */
+    DR_INSTR_CATEGORY_OTHER = 0x40         /* Other types of instructions */
```
What about instr_is_floating_ex() and dr_fp_type t (which seems to be completely missing for AArch64 today)?  I assume that would be implemented using the same category system: so we would want these types to also split into those categories?

kuhanov replied:
We will add on the next step. Need to discuss and fix that patch )))
Also we will share x86 approach. Uunfortunantly it is not programming just something like
```
[core/ir/x86/decode_table.c](https://github.com/DynamoRIO/dynamorio/commit/40cd2fa7cc53e54667b34430e8d5274bf713cecd#diff-f9b347cfa07cc46b1ea538b6e01652a0bdac4ad6c868183d734e4d823fa76853)
@@ -10703,4 +10703,1440 @@ const instr_info_t extra_operands[] =
    {OP_CONTD, 0xc00f0171, "<enclv cont'd>", ecx, edx, edx, xx, xx, mrm, x, END_LIST},
};

const uint category_by_opcode[] = {
    /* OP_INVALID */           DR_INSTR_CATEGORY_OTHER,
    /* OP_UNDECODED */         DR_INSTR_CATEGORY_OTHER,
    /* OP_CONTD */             DR_INSTR_CATEGORY_OTHER,
    /* OP_LABEL */             DR_INSTR_CATEGORY_OTHER,
    /* OP_add */               DR_INSTR_CATEGORY_MATH_INT,
   ...
```

derekbruening commented on core/ir/instr_api.h:
```diff
@@ -722,6 +724,11 @@ DR_API
 instr_get_opcode(instr_t *instr);
 
+DR_API
+/** Returns \p instr's set of categories (set of DR_INSTR_CATEGORY_ constants). */
```
I only see categories being set on decode.  What if I modify a decoded instruction: I don't see the categories being invalidated.

kuhanov replied:
Added check that `instr_operands_valid(instr))` is true. redecode if operands are invalid

derekbruening commented on core/CMakeLists.txt:
```diff
@@ -1209,6 +1209,7 @@ DR_export_header(${CMAKE_CURRENT_SOURCE_DIR}/ir/instrlist_api.h dr_ir_instrlist.
 DR_export_header(${CMAKE_CURRENT_SOURCE_DIR}/ir/dr_ir_opcodes.h dr_ir_opcodes.h)
 DR_export_header(${CMAKE_CURRENT_SOURCE_DIR}/ir/x86/opcode_api.h dr_ir_opcodes_x86.h)
 DR_export_header(${CMAKE_CURRENT_SOURCE_DIR}/ir/arm/opcode_api.h dr_ir_opcodes_arm.h)
+DR_export_header(${CMAKE_CURRENT_SOURCE_DIR}/ir/category_api.h dr_ir_categories.h)
```
Looks like this exported file will not be included by dr_api.h, unlike all the other files: that doesn't seem like what we want to suddenly require explicit includes for just this one file.

kuhanov replied:
Maybe we could remove this header at all? add categories to instr_api.h like dr_fp_type_t structure

derekbruening replied:
Yes that sounds fine to me.

derekbruening commented on core/ir/aarch64/codec.c:
```diff
@@ -9561,6 +9561,102 @@ encode_opnds_tbz(byte *pc, instr_t *instr, uint enc, decode_info_t *di)
+    /* Calculation of category is based on C4.1 'A64 instruction set encoding'
+     * of ARM V8 Architecture reference manual
+     *  https://developer.arm.com/documentation/ddi0487/
+     *  The encoding is:
```
Changes involving AArch64 encoding should be reviewed by @AssadHashmi

derekbruening commented:
> Hi, @derekbruening . Could you tell me if I need to make anything else? Could I merge this patch? Thx, Kirill

Sorry for the delay -- I went ahead and did a full review: mostly requesting documentation and tests.  It would also be best if @AssadHashmi takes a look for any AArch64 decoder changes.

We are definitely interested in this type of feature for other projects: #6238, as mentioned.  That is why we're trying to iterate on some of the category details to make sure the same interface will cover all uses.  Thanks for bearing with the review and discussion process.

kuhanov commented:
Hi, @derekbruening 
added the following:
redecode if instr_operands_valid(instr) is not valid; 
implement initial instr_is_floating_ex for AArch64
added test (is this the correct place where we added it?)
moved categories from separate header

Should we fix anything else?
Thx, Kirill

derekbruening commented on core/ir/instr_api.h:
```diff
@@ -287,7 +287,10 @@ struct _instr_t {
+    uint opcode : 16;
+
+    /* category of this instr (e.g. branch, load/store, etc.) */
+    uint category : 16;
```
This does seem like a good solution for your initial category set: I think shrinking the opcode will not cause any problems.  However, after some discussions here about categories we'd like to add on top of your initial ones, we're thinking a cap of 16 categories is too limiting.  We're about to bump the major version from 9 to 10 anyway so we're thinking let's go ahead and have a separate 32-bit category field and break compatibility now while the version bump is happening.

I'm seeing 3 32-bit fields at the top: if you add `uint category` after the `length` field it will not affect the 64-bit size (b/c it must currently have a padding slot to align the pointer field `bytes`) which will stay 112. 

This is still a compatibility break for 32-bit -- for that, update the black box as you did before, and the heap.c comment (I assume order didn't change there: asserts will fire if so), and should really bump the between-releases version too to satisfy our version policy https://dynamorio.org/page_new_release.html via: replace "9.93" with "9.94" in these files:
```
$ git grep '9\.93'
.github/workflows/ci-docs.yml:          export VERSION_NUMBER=9.93.$((`git log -n 1 --format=%ct` / (60*60*24)))
.github/workflows/ci-package.yml:          export VERSION_NUMBER=9.93.$((`git log -n 1 --format=%ct` / (60*60*24)))
.github/workflows/ci-package.yml:          export VERSION_NUMBER=9.93.$((`git log -n 1 --format=%ct` / (60*60*24)))
.github/workflows/ci-package.yml:          export VERSION_NUMBER=9.93.$((`git log -n 1 --format=%ct` / (60*60*24)))
.github/workflows/ci-package.yml:          export VERSION_NUMBER=9.93.$((`git log -n 1 --format=%ct` / (60*60*24)))
.github/workflows/ci-package.yml:          export VERSION_NUMBER=9.93.$((`git log -n 1 --format=%ct` / (60*60*24)))
.github/workflows/ci-package.yml:          export VERSION_NUMBER="9.93.$((`git log -n 1 --format=%ct` / (60*60*24)))"
CMakeLists.txt:set(VERSION_NUMBER_DEFAULT "9.93.${VERSION_NUMBER_PATCHLEVEL}")
```
Plus change `OLDEST_COMPATIBLE_VERSION_DEFAULT` to "994" in the top-level CMakeLists.txt.
(Or if PR #5835 goes in first and already bumps these to 9.94: bump further to 9.95.)

(Note that even with this change as-is to make `opcode` a 16-bit field we have compatibility breakage as the upper bits of opcode will now be set, even if no size change.)

kuhanov replied:
Done. Upgrade to 9.95 and add to changes list

derekbruening commented on core/lib/dr_api.h:
```diff
@@ -101,6 +101,8 @@ extern "C" {
 /* Annotation handler registration routines */
 #include "dr_annotation.h"
 
+/* Instruction category constants */
```
Looks like this should be removed.

derekbruening commented on core/ir/instr_shared.c:
```diff
@@ -449,6 +449,20 @@ instr_get_opcode(instr_t *instr)
+    (IF_DEBUG_(CLIENT_ASSERT(sizeof(*instr) == sizeof(instr_t), "invalid type"))( \
+        ((instr)->category == DR_INSTR_CATEGORY_UNCATEGORIZED ||                  \
+         !instr_operands_valid(instr))                                            \
+            ? (instr_decode_with_current_dcontext(instr), (instr)->category)      \
```
But as mentioned in the prior review round, what about a synthetic instr with no raw bits?  Does an encoding have to be done (cached in the raw bits) and then a decoding?  At the very least, document that this API is not supported for synthetic instructions and is only supported for decoded instructions (and someone can add synthetic support later: document the lack of support in the header, and here put a `XXX i#6238: ...` comment about it).

derekbruening commented on suite/tests/api/drdecode_aarch64.c:
```diff
@@ -130,6 +130,69 @@ test_noalloc(void)
+{
+    const uint raw[] = {
+        0x00000000, /* no category */
+        0x12020000, /* int math */
```
Add the disassembly string in the comment for each so a human can understand this.

derekbruening commented on core/ir/instr_api.h:
```diff
@@ -1854,6 +1875,20 @@ DR_API
+ * Indicates which category the instruction corresponds to.
+ */
+typedef enum {
+    DR_INSTR_CATEGORY_UNCATEGORIZED = 0x0, /**< Uncategorised. */
```
nit: Weird to use the American "z" in the value but British "s" in the comment.

derekbruening commented on suite/tests/api/drdecode_aarch64.c:
```diff
@@ -130,6 +130,69 @@ test_noalloc(void)
+    };
+
+    size_t instr_count = sizeof(raw) / sizeof(uint);
+    const uint categories[] = { DR_INSTR_CATEGORY_UNCATEGORIZED,
```
Would people expect _UNCATEGORIZED to mean an error: e.g., a synthetic instr for which we have no info?  And this should return _OTHER?

kuhanov replied:
We suppose that "UNCATEGORIZED" means we don't try to decode and set categories at all. _OTHERS means that we decode instruction and it doesn't assigned to known categories and move to othres

derekbruening commented on suite/tests/api/drdecode_aarch64.c:
```diff
@@ -130,6 +130,69 @@ test_noalloc(void)
+        ASSERT(cat == categories[i]);
+        pc += 4;
+    }
+}
```
Add a test of a synthetic instr in an obvious category (e.g., `INSTR_CREATE_ldr`) and confirm it returns _UNCATEGORIZED.

derekbruening reviewed (commented):
If you could actually put back the full 32-bit field, but move it to a better spot, that will help us add more categories later (see comment on that).  Mostly just have minor comments elsewhere; should be good to go after this round.  Thank you for working through the review rounds.

AssadHashmi commented on core/ir/aarch64/codec.c:
```diff
@@ -9561,6 +9561,106 @@ encode_opnds_tbz(byte *pc, instr_t *instr, uint enc, decode_info_t *di)
+    else if (BITS(enc, 31, 31) == 0 && op1 == 0) /* op1 is 0 and 31 bit is 0 */
+        category = DR_INSTR_CATEGORY_UNCATEGORIZED;
+    else {
+        /*                       op1 - xxxx
```
Might be worth adding a `DR_INSTR_CATEGORY_SVE`  with `op1` = `0010`. The codec supports all SVE instructions.

AssadHashmi replied:
Hi @kuhanov, are you handling `op1` = `0001` and `op1` = `x111` as being in the same category of SIMD vector instructions for the purposes of this PR?

Bear in mind that the Advanced SIMD (NEON) instructions are for _fixed width_ 128 bit vectors. The op1 = 0010 encoding class covers _scalable vector widths_ from 128 to 2048 bits depending on the hardware implementation. SVE is a separate encoding class from SIMD even though they both handle vectors. There are 337 SVE opcodes, not counting the latest SVE2 instructions.

derekbruening replied:
I would agree with @kuhanov on the cross-arch general _SIMD category grouping Neon and SVE for this PR.  I think the intended use cases for these categories are for more general purposes not tied to any one architecture.  Distinguishing individual ISA-specific characteristics does have its own uses cases but I think that would be more along the lines of FEATURE_ bits and would be a separate thing from these categories.

derekbruening commented:
(Please re-request review when ready)

kuhanov commented:
> t review when ready)

Hi, @derekbruening 
We are ready.  Could you approve? os should we fix anything else?
Thx, Kirill

derekbruening reviewed (approved):
Thanks, I will merge it.
